### PR TITLE
Update setFromMat4 and add tests

### PR DESCRIPTION
**Pull request by slimbuck**
Fixes: #3759

This PR:
* Updates the matrix to quaternion implementation to be more accurate (quaternions were drifting out of normalisation before).
* Add `Quat.mulScalar`
* Also add a suit of tests

### CODE REVIEW

File: src/core/math/quat.js [hunk 1]
```diff
@@ -461,103 +476,59 @@ class Quat {
...
+        } else {
+            if (m00 < -m11) {
+                this.set(m20 + m02, m12 + m21, 1 - m00 - m11 + m22, m01 - m10);
+            } else {
+                this.set(m12 - m21, m20 - m02, m01 - m10, 1 + m00 + m11 + m22);
             }
         }
 
-        return this;
+        return this.mulScalar(1.0 / this.length());
```
Comment (willeastcott) [inline src/core/math/quat.js:531]: Maybe put a comment about why your need to renormalize?

**Comment by willeastcott** [comment]
Any performance difference?

File: src/core/math/quat.js (hunk 1 above)
Comment (slimbuck) [reply src/core/math/quat.js:535]: done

**Comment by slimbuck** [comment]
> Any performance difference?

Performance difference should be negligible.

File: src/core/math/quat.js [hunk 2]
```diff
@@ -461,103 +476,63 @@ class Quat {
...
+                this.set(m12 - m21, m20 - m02, m01 - m10, 1 + m00 + m11 + m22);
             }
         }
 
-        return this;
+        // Instead of scaling by 0.5 / Math.sqrt(t) (to match the original implementation),
+        // instead we normalize the result. It costs 3 more adds and muls, but we get
+        // a stable result and in some cases normalization is required anyway, see
+        // https://github.com/blender/blender/blob/v4.1.1/source/blender/blenlib/intern/math_rotation.c#L368
+        return this.mulScalar(1.0 / this.length());
```
Comment (kungfooman) [inline src/core/math/quat.js:535]: Isn't this just what `Quat#normalize` is, e.g. `return this.normalize()`? Then we also don't need to extend the API by adding a new method.

Comment (slimbuck) [reply src/core/math/quat.js:535]: Yep, similar, except we don't need to test for length 0. I wasn't sure the effort here was necessary, but we decided adding `mulScalar` to match the other types is probably a win.

**Pull request merged**